File: restclient/restClient.go
package restclient

import (
	"fmt"
	"net/http"
	"sync"
	"time"
)

var testCounts = []int{100, 200, 300}

const baseURL = "http://localhost:5000"

var endpoints = map[string]string{
	"usuarios":  "/usuarios",
	"musicas":   "/musicas",
	"playlists": "/playlists",
}

func testLoad(endpoint, name string, numTest int) {
	url := baseURL + endpoint
	start := time.Now()

	var wg sync.WaitGroup
	rateLimiter := time.Tick(10 * time.Millisecond) // 100 requests per second

	for i := 0; i < numTest; i++ {
		<-rateLimiter
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			resp, err := http.Get(url)
			if err != nil {
				fmt.Printf("Request %d failed: %v\n", i, err)
				return
			}
			defer resp.Body.Close()
			if resp.StatusCode != http.StatusOK {
				fmt.Printf("Request %d failed with status: %s\n", i, resp.Status)
			}
		}(i)
	}
	wg.Wait()

	end := time.Now()
	fmt.Printf("%s load test with %d requests completed in %v seconds\n", name, numTest, end.Sub(start).Seconds())
}

func RunTests() {
	for _, numTest := range testCounts {
		fmt.Printf("Running tests with %d requests:\n", numTest)
		testLoad(endpoints["usuarios"], "Usuarios", numTest)
		testLoad(endpoints["musicas"], "Musicas", numTest)
		testLoad(endpoints["playlists"], "Playlists", numTest)
	}
}

File: python/clients/rest_client.py
import requests
import time

BASE_URL = "http://localhost:5000"
ENDPOINTS = ["/usuarios", "/musicas", "/playlists"]
TEST_COUNTS = [100, 200, 300]


def test_load(endpoint, name, num_test):
    url = f"{BASE_URL}{endpoint}"
    start = time.time()

    for i in range(num_test):
        try:
            response = requests.get(url)
            response.raise_for_status()
        except requests.RequestException as e:
            print(f"Request {i} failed: {e}")

    end = time.time()
    print(
        f"{name} load test with {num_test} requests completed in {end - start:.6f} seconds"
    )


def run_tests():
    for num_test in TEST_COUNTS:
        print(f"Running tests with {num_test} requests:")
        for endpoint in ENDPOINTS:
            test_load(endpoint, endpoint, num_test)


if __name__ == "__main__":
    run_tests()
